JitsiAuthConnection.js

```diff
@@ -0,0 +1,141 @@
+import * as JitsiConnectionEvents from './JitsiConnectionEvents';
+import XMPP from './modules/xmpp/xmpp';
+
+/**
+ * Class creates separate XMPP connection and tries to connect using given
+ * credentials. Once connected will contact Jicofo to obtain and store session
+ * ID which is then used by the parent conference to upgrade user's role to
+ * moderator. It's also used to join the conference when starting from anonymous
+ * domain and only authenticated users are allowed to create new rooms.
+ */
+export default class JitsiAuthConnection {
+    /**
+     * Creates new <tt>JitsiAuthConnection</tt> for given conference.
+     * @param {JitsiConference} conference
+     * @constructor
+     */
+    constructor(conference) {
+        this.conference = conference;
+        this.xmpp = new XMPP(conference.connection.options);
+        this.canceled = false;
+        this._promise = null;
+    }
+
+    /**
+     * @typedef {Object} UpgradeRoleError
+     * @property {JitsiConnectionErrors} [connectionError] - One of
+     * {@link JitsiConnectionErrors} which occurred when trying to connect to
+     * the XMPP server.
+     * @property {String} [authenticationError] - One of XMPP error conditions
+     * returned by Jicofo on authentication attempt. See
+     * https://xmpp.org/rfcs/rfc3920.html#streams-error.
+     * @property {String} [message] - More details about the error.
+     *
+     * NOTE If neither one of the errors is present it means that the operation
+     * has been canceled.
+     */
+    /**
+     * Connects to the XMPP server using given credentials and contacts Jicofo
+     * in order to obtain a session ID (which is then stored in the local
+     * storage). User's role of the parent conference will be upgraded to
+     * moderator (by Jicofo).
+     *
+     * @param {Object} options
+     * @param {string} options.id - XMPP user's ID (user@xmpp-server.com).
+     * @param {string} options.password - User's password used to login.
+     * @param {string} [options.roomPassword] - Room password required to join
+     * the MUC room.
+     * @param {Function} [options.onLoginSuccessful] - Callback called when
```

Comment:
Why is this callback needed? Is there code that needs to run between the login and the role upgrade?